src/memtester_parsing.py

```diff
@@ -0,0 +1,189 @@
+# (c) Open Compute Project
+# (c) Google LLC
+# (c) Igor Semenov (goshik92@gmail.com)
+# 
+# Use of this source code is governed by an MIT-style
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/MIT.
+
+from sly import Lexer, Parser
+from dataclasses import dataclass
+from typing import Union, Callable
+import re
+
+@dataclass
+class BadAddress:
+    addr: int
+    left_val: int
+    right_val: int
+    
+@dataclass
+class BadLine:
+    addr: int
+    
+@dataclass
+class OkResult:
+    pass
+
+@dataclass
+class Test:
+    name: str
+    result: Union[OkResult, list[Union[BadAddress, BadLine]]]
+    def passed(self) -> bool:
+        return type(self.result) is OkResult
+    
+@dataclass
+class Loop:
+    index: int
+    tests: list[Test]
+    def failed_tests(self) -> list[Test]:
+        return [t for t in self.tests if not t.passed()]
+
+@dataclass
+class MemtesterCallbacks:
+    # Called when the next output line is read from memtester
+    line_ready: Callable[[str], None] = lambda line: None
+
+    # Called when a lexing or parsing error occurs
+    parsing_error: Callable[[str], None] = lambda desc: None
+    
+    # Called when memtester version is read
+    version_ready: Callable[[str, bool], None] = lambda version, is_known: None
+    
+    # Called when all loops finish
+    run_ready: Callable[[int], None] = lambda failed_loop_count: None
+    
+    # Called when a loop of memory testing finish
+    loop_ready: Callable[[Loop], None] = lambda loop: None
+    
+    # Called when a test within a loop finishes
+    test_ready: Callable[[Test], None] = lambda test: None
+
+
+# https://sly.readthedocs.io/en/latest/sly.html#writing-a-lexer
+class MemtesterLexer(Lexer):
+    tokens = { VER, OK, FLR, LOOP, TEST, HEX }
+
+    TEST = (
+        r"Stuck\sAddress|Random\sValue|Compare\s(XOR|SUB|MUL|DIV|OR|AND)|"
+        r"Sequential\sIncrement|Solid\sBits|Block\sSequential|Checkerboard|"
+        r"Bit\sSpread|Bit\sFlip|Walking\s(Ones|Zeroes)|(8|16)-bit\sWrites"
+    )
+    
+    @_(r"Loop\s[0-9]+/[0-9]+:")
+    def LOOP(self, t):
```

Comment:
The `t` object in grammar rules is [dynamic](https://github.com/dabeaz/sly/blob/33d4f5afc04c9cae78bd7dc094e08badc738409e/src/sly/yacc.py#L104), so type annotations will not increase code readability. The object's properties and use scenarios are covered in the official SLY doc, but I'll probably try to touch on this topic if I get to write an [intro article](https://github.com/opencomputeproject/ocp-diag-memtester/issues/3).

Reply:
I just realized I misread this code citation. I thought the call was related to the parser, not lexer. `t` is always of the `Token` type. Will we benefit from type-annotating this fact?

Reply:
generally speaking, it would be ideal to have type hints everywhere possible. This gives the best chance of catching bugs (and autocomplete for lang servers), but also general code architecture (because when the types no longer make sense or are conflictual, it's a clear sign that the code design has a problem -- this is a hint in python, other langs cover this by being compiled)

Reply:
>because when the types no longer make sense or are conflictual, it's a clear sign that the code design has a problem

I prefer statically typed languages over dynamically typed ones for that very reason. However, in this particular case, it's the SLY library who calls those methods, so I fail to see what design problems can be caught by type hinting. I do, however, agree with the autocompletion argument. Done.

UPD: No, not done. `Token` is not importable ([1](https://github.com/dabeaz/sly/blob/33d4f5afc04c9cae78bd7dc094e08badc738409e/src/sly/lex.py#L34), [2](https://github.com/dabeaz/sly/blob/33d4f5afc04c9cae78bd7dc094e08badc738409e/src/sly/__init__.py#L2)).

Reply:
heh.. sad when the lib you wanna use doesnt have typings